pkg/controller/storagecluster/noobaa_system_reconciler_test.go

```diff
@@ -395,3 +395,104 @@ func getReconciler(t *testing.T, objs ...runtime.Object) ReconcileStorageCluster
 		platform: &Platform{},
 	}
 }
+
+type testKMSArgs struct {
+	kmsProvider string
+	kmsAddress  string
+	testLabel   string
+	valid       bool
+}
+
+func TestNoobaaSystemKMS(t *testing.T) {
+	cr := createDefaultStorageCluster()
+	// KMS enabled
+	cr.Spec.Encryption.KeyManagementService.Enable = true
+	request := reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Name:      "ocsinit",
+			Namespace: "",
+		},
+	}
+	var _ = request
+	validKMSArgs := []testKMSArgs{
+		{testLabel: "case 1", kmsProvider: "vault", kmsAddress: "http://localhost:5050", valid: true},
+		{testLabel: "case 2", kmsProvider: "vault", kmsAddress: "http://localhost:30303", valid: true},
+		// ocs-operator is agnostic to KMS Provider, here rook should be throwing error
+		{testLabel: "case 3", kmsProvider: "newKMSProvider", kmsAddress: "http://127.0.0.1:1553", valid: true},
+		// invalid test cases, make sure label has a prefix 'invalid'
+		{testLabel: "invalid-case 1", kmsProvider: "vault", kmsAddress: "http://unearchable.url.location:3366", valid: false},
+	}
+	for _, kmsArgs := range validKMSArgs {
+		assertNoobaaKMSConfig(t, &kmsArgs, cr)
+	}
+}
+
+func assertNoobaaKMSConfig(t *testing.T, kmsArgs *testKMSArgs, cr *v1.StorageCluster) {
+	ctxTodo := context.TODO()
+	noobaaObj := &v1alpha1.NooBaa{}
+	noobaaObj.Name = "noobaa"
+	kmsCM := createDummyKMSConfigMap(kmsArgs.kmsProvider, kmsArgs.kmsAddress)
+
+	reconciler := createFakeInitializationStorageClusterReconciler(t, noobaaObj)
+	if err := reconciler.client.Create(ctxTodo, kmsCM); err != nil {
+		t.Errorf("Unable to create KMS configmap: %w", err)
+		t.FailNow()
```

Comment:
Why is this test using this pattern of Errorf and FailNow? Isn't that the entire reason we have assert?

Reply:
`assert.<functions>` will fall through even when the assert fails.

```go
func TestAsserts(t *testing.T) {
	err := fmt.Errorf("This is an error")
	assert.NoErrorf(t, err, "Not expecting any error, but received: %v", err)
	// if assert worked as expected, we should not reach the below statements
	err = nil
	assert.Errorf(t, err, "Expecting an error, but received: %v", err)
}
```

Output
```
$ go test -v -run "TestAsserts*"     
                                                                                                
=== RUN   TestAsserts
tmp_testgo:7:                                                                                               
                Error Trace:    tmp_test.go:7                                                                    
                Error:          Received unexpected error:                                                                                                              This is an error                                                                                        
                Test:           TestAsserts                         
                Messages:       Not expecting any error, but received: This is an error
tmp_test.go:9:                           
                Error Trace:    tmp_test.go:9 
                Error:          An error is expected but got nil.                                                                                       Test:           TestAsserts          
                Messages:       Expecting an error, but received: <nil>                                                                 
--- FAIL: TestAsserts (0.00s)                                                                                                           
FAIL                                                                                                                                    
exit status 1                                                                                                                           
my_assert_tests 0.037s
```

It is ok to provide assert functions  to assess different failures of the `main component` (continue after each failure and check the next functionality). But what happens if creation of `main component` itself fails? No point in assert through it's functionality failures now (as the component itself is not there).
So I had to use `FailNow()` function in places where the main component itself is not created (or failed to create); ie creation of `CephCluster` resource fails, no point in going further down to check the failures in `ensureNoobaaSystem()` functionality or to check KMS config values.